### DISCUSSION TITLE
newb

### DISCUSSION THREAD

**Discussion by Nighttripper501** [tomshaffner/PiThermalCam#10] [Q&A] [answered]
I am not able to get this to run?  Total newb at this.  This is what I've done so far.

```
sudo apt-get update &&  sudo apt-get upgrade

sudo apt-get install libatlas-base-dev
sudo apt-get install python-smbus
sudo apt-get install i2c-tools

sudo nano /boot/config.txt
# find the line with dtparam=i2c_arm=on. Add 
# "i2c_arm_baudrate=400000"
# after it

sudo i2cdetect -y 1
# You should see 33 in colum 3, row 30

PATH=$PATH:DIR/home/pi/.local/bin

pip3 install -r pithermalcam
or this after downloading the code.
pip3 install -r https://github.com/tomshaffner/PiThermalCam.git/requirements.txt

```
It's not making the folder in home.  So I made it and moved the code into it.  I also made the icons executionable.

**Comment by tomshaffner** [comment, marked as answer]
What happens when you try to run it?

**Reply by Nighttripper501** [reply]
Flask Thermal Camera - Pointer gets an hour glass for about 20 seconds.  That's it.
OpenCV Thermal Camera - Seems to work well
Matplotlib Thermal Camera -  Opens a terminal labeled "Python3" then writes a line of data then closes before I can read it.

**Reply by Nighttripper501** [reply]
I got it to work in server mode.  But it keeps switching between two maps.  I ran this command "sudo chmod +x web_server.py" and that opened Thonny.  Then I hit "run"


```
Server can be found at 192.168.1.10:8000
 * Serving Flask app "web_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
Traceback (most recent call last):
  File "/home/pi/PiThermalCam/pithermcam/web_server.py", line 153, in <module>
    start_server()
  File "/home/pi/PiThermalCam/pithermcam/web_server.py", line 148, in start_server
    app.run(host=ip, port=port, debug=False,threaded=True, use_reloader=False)
  File "/usr/lib/python3/dist-packages/flask/app.py", line 943, in run
    run_simple(host, port, self, **options)
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 814, in run_simple
    inner()
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 774, in inner
    fd=fd)
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 660, in make_server
    passthrough_errors, ssl_context, fd=fd)
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 577, in __init__
    self.address_family), handler)
  File "/usr/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/usr/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 98] Address already in use
```

Thank you for the awesome thermal cam
Mark

**Reply by tomshaffner** [reply]
You've seen the page that discusses the setup, right? https://tomshaffner.github.io/PiThermalCam/ has a LOT of information. Sounds, from the above, like the sections that discuss the various modes of running are worth digging into. E.g. writing one line to the screen and then stopping sounds like the mode for verification that you're getting data. See, for example the section on Matplotlib running modes: https://tomshaffner.github.io/PiThermalCam/#matplotlib-version.

Also, if the opencv version works, is that not all you need?

The error above sounds like the device is already being used. You might reboot the device and then try the web server version first to see if it works.

**Reply by Nighttripper501** [reply]
It does have a ton of information. But most is over my head.  I have been able to track down some of the questions I had by google.  When you say "Run pithermalcam/web_server.py to set up a flask server and stream live video over the local network.  How?  If I click the icon I moved to the desktop and made execution-able.  Nothing happens.  I made web_server.py executionable and when I click on it Thonny opens.  If I click run then the sever starts.  But the image switches between two different image styles.  Is there a way for the sever to run on startup?  With no keyboard or screen attached?  So all I would need is the pi and a device to view the webpage on.  Like a phone or a fire tablet?

Thank you
Mark

**Reply by tomshaffner** [reply]
Hi Mark,

Running in python means either starting a python environment (documentation at https://docs.python.org/3/tutorial/venv.html about one type of python environment) and running the script (by typing it's location/name), or typing something like `python3 pithermalcam/web_server.py` on the pi, which feeds the script directly to the python 3 instance. For more help with this you can search for anything related to how to run a python 3 script in linux.

Unfortunately, doing something like having the script run at boot time is going to require a fair bit of that kind of general python knowledge and some general linux knowledge also, and as such is outside the scope of support for this library. The paragraph above should help get you started running the code at least, and beyond that I'd suggest that the terms you're searching for are going to be focusing on running a python (python 3 to be specific, in case you find any old pages) script automatically at boot on a pi. You'll then want to target that process on the web server script.

Hope that helps!

**Comment by Nighttripper501** [comment]
Yes it does help, thank you.  This will give this 70 year old something new
to learn on these hot days ahead.  I really do appreciate you taking time
out to help me.

Thanks
Mark

**Comment by Nighttripper501** [comment]
I got your program running.  Learned a bit along the way.  I do have
something I haven't figured out.  The screen image keeps switching between
different views.  It does it on the pi and when I remotely connect.  It is
also passed on to the web server image.  This is a screen capture of the
video.  Any help would be very much appreciated.

http://www.burgus.me/mark/screen.webm

Thank you
Mark

**Comment by tomshaffner** [comment]
Hi Mark,
That's a bit strange. The first thing I'm wondering is if that's actually different views, or some underlying data/refresh issue. One easy way to test that would be to change to a different colormap and see if something similar happens. E.g. if you pick a different color scheme, does the same dynamic occur?
If so then it's not actually the view changing but something with the data itself potentially. Also of note to me, in the video at least the refresh rate seems quite low. Is this on a Pi? If so, have you changed the baudrate already? Instructions on this in the setup section at https://tomshaffner.github.io/PiThermalCam/#prerequisitesetup.
Let me know how that goes; hope it helps!
          Tom
    On Tuesday, July 6, 2021, 12:29:07 PM EDT, Nighttripper501 ***@***.***> wrote:  
 
 

I got your program running. Learned a bit along the way. I do have
something I haven't figured out. The screen image keeps switching between
different views. It does it on the pi and when I remotely connect. It is
also passed on to the web server image. This is a screen capture of the
video. Any help would be very much appreciated.

http://www.burgus.me/mark/screen.webm

Thank you
Mark

**Comment by Nighttripper501** [comment]
This is what I am doing to install your program.

"
sudo apt-get update &&  sudo apt-get upgrade
sudo apt-get install libatlas-base-dev
sudo nano /boot/config.txt
// find the line with dtparam=i2c_arm=on. Uncomment it and add
// "i2c_arm_baudrate=400000"
// after it
reboot
sudo i2cdetect -y 1
pip3 install pithermalcam
//
// The scripts f2py, f2py3 and f2py3.7 are installed in
'/home/pi/.local/bin' which is not on PATH.
//  Consider adding this directory to PATH or, if you prefer to suppress
this warning, use --no-warn-script-location.

sudo nano /home/pi/.local/lib/python3.7/site-packages/adafruit_mlx90640.py
// change 4 to 40 in line 154
python3 /home/pi/PiThermalCam/examples/web_server.py

**Comment by Nighttripper501** [comment]
I'm sorry I hit the wrong key and sent the email early.

I changed the 4 to a 40 because the program errored out.  With the change
it works.  So there is something wrong I'm doing.  I was using two
different pi 3's  But on this run I used a pi 4  Each one gives the same
results.  One thing I left out is I downloaded the file package from github
and placed the folder "PiThermalCam" in the /home/pi directory.  One other
odd thing.  When I run the program on the pi.  If I try to save a
screenshot it is not saved.  I looked around and found the save folder is
/home/pi/pithermalcam. No caps

Thank you for helping me out
Mark

**Comment by Nighttripper501** [comment]
I finally beat my head on the wall so much that a hole appeared and the
solution was on the other side.  Your program was telling me the problem Ii
just wasn't listening.  After installing "pip3 install pithermalcam"
I get this error message
"The scripts f2py, f2py3 and f2py3.7 are installed in '/home/pi/.local/bin'
which is not on PATH.
Consider adding this directory to PATH or, if you prefer to suppress this
warning, use --no-warn-script-location."
Duh.  I used the cmd
export PATH="$HOME/.local/bin:$PATH"
And your program works awesome.  If only I would learn to listen more.

Thank you for the awesome program and your many many hours of programming.
Mark

**Reply by tomshaffner** [reply]
Congrats on the breakthrough!

I make that kind of mistake all the time. It's amazing how often I ignore the warning message right in front of me until I've checked every other irrelevant possibility out there. Glad to know I'm not the only programmer that does that. 😄 Thanks for sticking with it and posting the result!
